lib/puppet/provider/agent_upgrade/windows.rb

```diff
@@ -0,0 +1,85 @@
+
+Puppet::Type.type(:agent_upgrade).provide(:agent_upgrade) do
```

Comment:
Since `puppet-agent` is a package, and can already be managed using the built-in `windows` package provider, I think it is confusing to have a type & provider that do a similar thing. For example, the instances method below returns `puppet-agent` packages.

Another option would be to create a class that:

1. Validates the parameters, e.g. source URL, arch, ...
1. Uses the powershell module to execute powershell code passing in the current ruby process id.
1. The powershell code spawns another detached powershell instance that waits for the ruby process to exit, and then executes msiexec